DOCSP-43348 -- DB and Application Resilience (#26)
mongodb/docs-atlas-architecture: source/resiliency.txt

>    :depth: 1

Reviewer: ```suggestion
   :depth: 2
```

   :class: onecol

> MongoDB |service| is a modern, highly-performant database that is purpose-built 
> to maintain uptime regardless of infrastructure outages, system maintenance and more.

Reviewer: Just tweaking this slightly since some of the language feels a little market-y, and to lengthen it
```suggestion
MongoDB |service| is a highly-performant database that is designed 
to maintain uptime regardless of infrastructure outages, system maintenance, and more. Use the guidance on this page to plan settings to maximize the resiliency of your application and database.
```

> {+service+} Features and Best Practices for Resiliency
> ------------------------------------------------------

Reviewer: ```suggestion
{+service+} Features and Recommendations for Resiliency
-------------------------------------------------------
```

> Database Replication
> ~~~~~~~~~~~~~~~~~~~~
> 

Reviewer: Changing this for parity with other pages. Can you adjust the rest of the heading levels below this as well please?
```suggestion
Features
~~~~~~~

Database Replication
````````````````````

```

> |service| {+clusters+} consist of a minimum of three nodes and can be increased 

Reviewer: Passive voice
```suggestion
|service| {+clusters+} consist of a minimum of three nodes, and you can increase the node count 
```

to any odd number of nodes you require. Data from your application is first written 
to a primary node, and that data is then incrementally replicated and stored on 
all secondary nodes within your cluster.

By default, {+cluster+} nodes are distributed across availability zones within 
one of your chosen cloud provider's availability regions. For example, if your 
cluster is deployed to the cloud provider region ``us-east``, |service| deploys 

> nodes to ``us-east-a``, ``us-east-b`` and ``us-east-c`` by default. 

Reviewer: ```suggestion
nodes to ``us-east-a``, ``us-east-b`` and ``us-east-c`` by default. 

To learn more about high availability and node distribution across regions, see :ref:`arch-center-high-availability`.
```

Self-Healing Deployments
````````````````````````

|service| {+clusters+} must consist of an odd number of nodes, because only one 
node can be elected as the primary node to and from which your application writes 
and reads directly. 

In the event that a primary node is unavailable, because of infrastructure 
outages, maintenance windows or any other reason, |service| {+clusters+} self-heal by 
converting an existing secondary node into your primary node to maintain 
database availability. 

Maintenance Window Uptime
`````````````````````````

|service| maintains uptime during scheduled maintenance by applying updates in 
a rolling fashion to one node at a time. During this process, |service| elects a new 
primary when necessary just as it does during any other unplanned primary node 
outage.  

Monitoring
~~~~~~~~~~

|service| provides `built-in tools <https://www.mongodb.com/docs/atlas/monitoring-alerts/>`__ to monitor cluster performance, query performance 

> and more. Additionally, |service| integrates easily with `third-party services <https://www.mongodb.com/docs/atlas/tutorial/third-party-service-integrations/#std-label-third-party-integrations>`__. 

Reviewer: Can we add:

- A sentence on why this matters (if you have visibility into cluster performance issues, you can address issues quickly)
- A link to :ref:`arch-center-monitoring-alerts` to learn more

> Building Resilient Applications with {+service+}
> ------------------------------------------------

Reviewer: To align with other pages, please adjust other heading levels below this
```suggestion
Recommendations
~~~~~~~~~~~~~~
```

`Query timeout <https://www.mongodb.com/docs/manual/tutorial/query-documents/specify-query-timeout/>`__ behavior can be set globally in |service| and it can be defined 
at the query level. 

Retryable Database Reads and Writes
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

|service| supports `retryable read <https://www.mongodb.com/docs/manual/core/retryable-reads/>`__ and `retryable write <https://www.mongodb.com/docs/manual/core/retryable-writes/>`__ operations. When enabled, 
read and write operations are retried once as a safeguard against intermittent 
network outages. 

Configure Read and Write Concerns 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

|service| {+clusters+} eventually replicate all data across all nodes. However, 
you can configure the number of nodes across which data must be repicated before 
a read or write operation is reported to have been successful. `Read concerns <https://www.mongodb.com/docs/manual/reference/read-concern/>`__ and 
`write concerns <https://www.mongodb.com/docs/manual/reference/write-concern/>`__ can be defined globally in |service| and they can be defined at the 
client level in the connection string. 

Examples

> --------

Reviewer: If they can't set these things in the CLI or TF, let's remove the example section from this page.

Author: I went ahead and removed for now, as I can't yet access dev clusters with working API keys, and I think a description and link out to the existing docs works here. Happy to adjust as needed. Thanks
